src/main/java/org/cbioportal/service/impl/ClinicalDataDensityPlotServiceImpl.java

```diff
@@ -0,0 +1,163 @@
+                densityPlotBin.setBinY(BigDecimal.valueOf(yAxisStartValue + (j * yAxisBinInterval)));
+                densityPlotBin.setCount(0);
+                bins.add(densityPlotBin);
+            }
+        }
+
+        for (int i = 0; i < xValues.length; i++) {
+            double xValue = xValues[i];
+            double yValue = yValues[i];
+            int xBinIndex = (int) ((xValue - xAxisStartValue) / xAxisBinInterval);
+            int yBinIndex = (int) ((yValue - yAxisStartValue) / yAxisBinInterval);
+            int index = (int) (((xBinIndex - (xBinIndex == xAxisBinCount ? 1 : 0)) * yAxisBinCount) +
+                (yBinIndex - (yBinIndex == yAxisBinCount ? 1 : 0)));
+            DensityPlotBin densityPlotBin = bins.get(index);
+            densityPlotBin.setCount(densityPlotBin.getCount() + 1);
+            BigDecimal xValueBigDecimal = BigDecimal.valueOf(xValue);
+            BigDecimal yValueBigDecimal = BigDecimal.valueOf(yValue);
+            if (densityPlotBin.getMinX() != null) {
+                if (densityPlotBin.getMinX().compareTo(xValueBigDecimal) > 0) {
+                    densityPlotBin.setMinX(xValueBigDecimal);
+                }
+            } else {
+                densityPlotBin.setMinX(xValueBigDecimal);
+            }
+            if (densityPlotBin.getMaxX() != null) {
+                if (densityPlotBin.getMaxX().compareTo(xValueBigDecimal) < 0) {
+                    densityPlotBin.setMaxX(xValueBigDecimal);
+                }
+            } else {
+                densityPlotBin.setMaxX(xValueBigDecimal);
+            }
+            if (densityPlotBin.getMinY() != null) {
+                if (densityPlotBin.getMinY().compareTo(yValueBigDecimal) > 0) {
+                    densityPlotBin.setMinY(yValueBigDecimal);
+                }
+            } else {
+                densityPlotBin.setMinY(yValueBigDecimal);
+            }
+            if (densityPlotBin.getMaxY() != null) {
+                if (densityPlotBin.getMaxY().compareTo(yValueBigDecimal) < 0) {
+                    densityPlotBin.setMaxY(yValueBigDecimal);
+                }
+            } else {
+                densityPlotBin.setMaxY(yValueBigDecimal);
+            }
+        }
+
+        if (xValues.length > 1) {
+            // need at least 2 entries in each to compute correlation
+            result.setPearsonCorr(new PearsonsCorrelation().correlation(xValues, yValues));
+            result.setSpearmanCorr(new SpearmansCorrelation().correlation(xValues, yValues));
+        } else {
+            // if less than 1 entry, just set 0 correlation
+            result.setSpearmanCorr(0.0);
+            result.setPearsonCorr(0.0);
+        }
+
+        // filter out empty bins
+        result.setBins(result.getBins().stream().filter((bin)->(bin.getCount() > 0)).collect(Collectors.toList()));
+        return result;
+    }
+
+    @Override
+    public List<ClinicalData> filterClinicalData(List<ClinicalData> clinicalDataList) {
```

Comment:
not sure if i understand, this is not yet wired up with clinical filters, right?

Reply:
I used the wrong name, hence your confusion. This basically extracts only values that can be used in the density plot. It just removes samples that don't have data for botht he X and Y axis, as well as removes non-numeric ClinicalData.